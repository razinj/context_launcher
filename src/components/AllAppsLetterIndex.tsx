import React from 'react'
import {
  FlatList,
  ListRenderItem,
  ListRenderItemInfo,
  PressableAndroidRippleConfig,
  StyleSheet,
  Text,
  View,
} from 'react-native'
import { useSelector } from 'react-redux'
import { WHITE_COLOR } from '../constants'
import { AppLetterIndex } from '../models/list-letter-index'
import { selectAppsLetterListMemoized } from '../slices/appsList'
import CustomPressable from './shared/CustomPressable'

type Props = {
  onPress: (letterIndex: number) => void
}

const rippleConfig: PressableAndroidRippleConfig = {
  borderless: false,
  foreground: true,
  color: '#e5e5e5',
  radius: 12,
}

const keyExtractor = ({ letter }: AppLetterIndex) => letter

const AllAppsLetterIndex = ({ onPress }: Props) => {
  const appsLetterList = useSelector(selectAppsLetterListMemoized)

  const renderItem: ListRenderItem<AppLetterIndex> = ({ item }: ListRenderItemInfo<AppLetterIndex>) => {
    return (
      <CustomPressable
        android_ripple={rippleConfig}
        style={styles.letterIndexLabelWrapper}
        onPress={() => onPress(item.index)}>
        <Text style={styles.letterIndexLabel}>{item.letter}</Text>
      </CustomPressable>
    )
  }

  return (
    <View style={styles.letterIndexContainer} testID='all-apps-letter-index-wrapper'>
      <FlatList
        testID='all-apps-letter-index-list'
        data={appsLetterList}
        renderItem={renderItem}
        keyExtractor={keyExtractor}
        initialNumToRender={appsLetterList.length}
      />
    </View>
  )
}

const styles = StyleSheet.create({
  letterIndexContainer: {
    top: 0,
    right: 0,
    bottom: 0,
    position: 'absolute',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  letterIndexLabelWrapper: {
    width: 25,
    height: 25,
    alignItems: 'center',
    justifyContent: 'center',
  },
  letterIndexLabel: {
    color: WHITE_COLOR,
    fontSize: 12,
    fontWeight: '600',
    textShadowColor: 'rgba(0, 0, 0, 0.75)',
    textShadowRadius: 2.5,
  },
})

export default AllAppsLetterIndex
