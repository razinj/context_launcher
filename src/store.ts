import AsyncStorage from '@react-native-async-storage/async-storage'
import { combineReducers, configureStore } from '@reduxjs/toolkit'
import {
  createMigrate,
  FLUSH,
  MigrationManifest,
  PAUSE,
  PERSIST,
  PersistConfig,
  persistReducer,
  persistStore,
  PURGE,
  REGISTER,
  REHYDRATE,
} from 'redux-persist'
import createSagaMiddleware from 'redux-saga'
import rootSaga from './rootSaga'
import appsListReducer from './slices/appsList'
import appStateReducer from './slices/appState'
import FavoriteAppsReducer from './slices/favoriteApps'
import pinnedAppsReducer from './slices/pinnedApps'
import preferencesReducer from './slices/preferences'
import recentAppsReducer from './slices/recentApps'

export const rootReducer = combineReducers({
  appsList: appsListReducer,
  recentApps: recentAppsReducer,
  favoriteApps: FavoriteAppsReducer,
  preferences: preferencesReducer,
  pinnedApps: pinnedAppsReducer,
  appState: appStateReducer,
})

const migrations: MigrationManifest = {
  2: (state: any) => {
    return {
      ...state,
      preferences: {
        displayRecentApps: true,
        displayFavoriteApps: true,
      },
    }
  },
  3: (state: any) => {
    return {
      ...state,
      pinnedApps: {
        list: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      preferences: {
        ...state.preferences,
        displayPinnedApps: true,
        displayTemporaryPinnedApps: false,
      },
    }
  },
  4: (state: any) => {
    const localState = { ...state }
    delete localState.appsSearch

    return {
      ...localState,
      appsList: {
        list: [],
      },
      pinnedApps: {
        list: [],
        temporarily: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      appState: {
        displayAllApps: false,
        displaySettings: false,
        menuAppDetails: undefined,
        displayAppMenu: false,
        displaySortableFavoriteApps: false,
        displaySortablePinnedApps: false,
        search: {
          query: undefined,
          result: [],
        },
      },
    }
  },
}

const persistConfig: PersistConfig<any> = {
  key: 'root',
  debug: false,
  version: 4,
  storage: AsyncStorage,
  blacklist: ['appState'],
  migrate: createMigrate(migrations, { debug: false }),
}

const persistedReducer = persistReducer(persistConfig, rootReducer)

const sagaMiddleware = createSagaMiddleware()

export const store = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat(sagaMiddleware),
})

sagaMiddleware.run(rootSaga)

export const persistor = persistStore(store)

export type RootState = ReturnType<typeof rootReducer>
export type AppDispatch = typeof store.dispatch
