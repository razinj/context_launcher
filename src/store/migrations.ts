import { MigrationManifest } from 'redux-persist'

export const migrations: MigrationManifest = {
  2: (state: any) => {
    return {
      ...state,
      preferences: {
        displayRecentApps: true,
        displayFavoriteApps: true,
      },
    }
  },
  3: (state: any) => {
    return {
      ...state,
      pinnedApps: {
        list: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      preferences: {
        ...state.preferences,
        displayPinnedApps: true,
        displayTemporaryPinnedApps: false,
      },
    }
  },
  4: (state: any) => {
    const localState = { ...state }
    delete localState.appsSearch

    return {
      ...localState,
      appsList: {
        list: [],
      },
      pinnedApps: {
        list: [],
        temporarily: [],
        temporaryAppsConfig: {
          startDate: undefined,
          endDate: undefined,
        },
      },
      appState: {
        displayAllApps: false,
        displaySettings: false,
        menuAppDetails: undefined,
        displayAppMenu: false,
        displaySortableFavoriteApps: false,
        displaySortablePinnedApps: false,
        search: {
          query: undefined,
          result: [],
        },
      },
    }
  },
}
