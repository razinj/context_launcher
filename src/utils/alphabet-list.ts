import { AppDetails } from '../models/app-details'
import { AppLetterIndex } from '../models/list-letter-index'

type LetterMap = {
  [key: string]: number
}

const alphabet = [
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
]

/**
 * @returns Indexed alphabet letters in an object where the key is the alphabet and the value is the index
 */
export const getLettersMap = (): LetterMap => {
  const letterMap: LetterMap = {}

  alphabet.forEach((letter: string, index: number) => (letterMap[letter] = index + 1))

  return letterMap
}

/**
 * @param value Any string value
 * @returns A letter if the first chareceter of the value is an alphabet, otherwise it returns a '#'
 */
export const getFirstLetter = (value: string): string => {
  const firstChar = value.trim().substring(0, 1).toUpperCase()
  const isValidLetter = getLettersMap()[firstChar]

  if (isValidLetter) return firstChar

  return '#'
}

export const getAppsLetterIndex = (apps: AppDetails[]): AppLetterIndex[] => {
  const appsLetterIndex: AppLetterIndex[] = []
  const treatedLetters: string[] = []

  apps.forEach(({ name }: AppDetails, index: number) => {
    const letter = getFirstLetter(name)

    if (treatedLetters.includes(letter)) return

    appsLetterIndex.push({ letter, index })
    treatedLetters.push(letter)
  })

  return appsLetterIndex
}
