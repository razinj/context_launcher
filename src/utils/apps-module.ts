// Native modules
import AppsModule from '../native-modules/AppsModule'

export const launchApp = (packageName: string) => AppsModule.launchApplication(packageName)

export const showAppDetails = (packageName: string) => AppsModule.showApplicationDetails(packageName)

export const requestAppUninstall = (packageName: string) => AppsModule.requestApplicationUninstall(packageName)
