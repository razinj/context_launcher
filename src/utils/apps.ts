import { APP_ITEM_HEIGHT } from '../constants'
import { AppDetails } from '../models/app-details'
import { PinnedApp } from '../models/pinned-app'

export const getAppsByLabel = (apps: AppDetails[], query: string): AppDetails[] => {
  return apps.filter((app: AppDetails) => app.name.match(new RegExp(query, 'gi')))
}

export const getAppIndex = (apps: AppDetails[], appPackageName: string): number => {
  return apps.findIndex(({ packageName }: AppDetails) => packageName === appPackageName)
}

export const getApp = (apps: AppDetails[], appPackageName: string): AppDetails | PinnedApp | undefined => {
  return apps.find(({ packageName }: AppDetails) => packageName === appPackageName)
}

export const getListKey = ({ packageName }: AppDetails) => packageName

export const getListItemLayout = (_data: unknown, index: number) => ({
  length: APP_ITEM_HEIGHT,
  offset: APP_ITEM_HEIGHT * index,
  index,
})
