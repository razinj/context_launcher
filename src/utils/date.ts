export const getTimeFromDate = (stringDate?: string): string => {
  if (!stringDate) return 'unset'

  const date = new Date(stringDate)

  const singleDigitHours = `${date.getHours()}`.length == 1
  const singleDigitMinutes = `${date.getMinutes()}`.length == 1

  const hours = singleDigitHours ? `0${date.getHours()}` : date.getHours()
  const minutes = singleDigitMinutes ? `0${date.getMinutes()}` : date.getMinutes()

  return `${hours}:${minutes}`
}

export const getDateFromStringWithCurrentDateValue = (stringDate?: string): Date | undefined => {
  if (!stringDate) return

  const currentDate = new Date()
  const fullDate = new Date(stringDate)
  return new Date(
    currentDate.getFullYear(),
    currentDate.getMonth(),
    currentDate.getDate(),
    fullDate.getHours(),
    fullDate.getMinutes(),
    0,
    0
  )
}

export const stripDateFromSeconds = (date: Date): Date => {
  return new Date(date.getFullYear(), date.getMonth(), date.getDate(), date.getHours(), date.getMinutes(), 0, 0)
}
