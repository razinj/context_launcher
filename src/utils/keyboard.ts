import { Keyboard } from 'react-native'

export const dismissKeyboard = () => Keyboard.dismiss()
