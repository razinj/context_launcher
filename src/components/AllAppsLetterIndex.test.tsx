import { fireEvent, screen } from '@testing-library/react-native'
import React from 'react'
import { getAppForTestsByName, initialStoreState } from '../../utils/test/data'
import { renderWithProvider } from '../../utils/test/utils'
import { RootState } from '../store'
import AllAppsLetterIndex from './AllAppsLetterIndex'

describe('<AllAppsLetterIndex /> Tests', () => {
  const onPressFn = jest.fn()

  it('should render correctly and match snapshot - empty list', () => {
    renderWithProvider(<AllAppsLetterIndex onPress={onPressFn} />)

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('all-apps-letter-index-wrapper')).toBeOnTheScreen()
    expect(screen.getByTestId('all-apps-letter-index-list')).toBeOnTheScreen()
  })

  it('should render correctly and match snapshot - populated list', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      appsList: {
        list: [getAppForTestsByName('Chrome')!],
        loading: false,
      },
    }

    renderWithProvider(<AllAppsLetterIndex onPress={onPressFn} />, { preloadedState: customInitialState })

    expect(screen.toJSON()).toMatchSnapshot()
    expect(screen.getByTestId('all-apps-letter-index-wrapper')).toBeOnTheScreen()

    const list = screen.getByTestId('all-apps-letter-index-list')

    expect(list).toBeOnTheScreen()

    const letterIndexItem = screen.getByText('C')

    expect(letterIndexItem).toBeOnTheScreen()
    expect(list).toContainElement(letterIndexItem)
  })

  it('should call passed function when element is pressed', () => {
    const customInitialState: RootState = {
      ...initialStoreState,
      appsList: {
        list: [getAppForTestsByName('Chrome')!, getAppForTestsByName('Maps')!],
        loading: false,
      },
    }

    renderWithProvider(<AllAppsLetterIndex onPress={onPressFn} />, { preloadedState: customInitialState })

    const list = screen.getByTestId('all-apps-letter-index-list')

    expect(list).toBeOnTheScreen()

    // Letter C
    const letterIndexItem_C = screen.getByText('C')

    expect(letterIndexItem_C).toBeOnTheScreen()
    expect(list).toContainElement(letterIndexItem_C)

    fireEvent.press(letterIndexItem_C)

    expect(onPressFn).toHaveBeenCalledWith(0)

    // Letter M
    const letterIndexItem_M = screen.getByText('M')

    expect(letterIndexItem_M).toBeOnTheScreen()
    expect(list).toContainElement(letterIndexItem_M)

    fireEvent.press(letterIndexItem_M)

    expect(onPressFn).toHaveBeenCalledWith(1)
  })
})
