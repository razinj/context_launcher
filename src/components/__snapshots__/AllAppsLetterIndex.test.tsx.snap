// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AllAppsLetterIndex /> Tests should render correctly and match snapshot - empty list 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="all-apps-letter-index-wrapper"
    >
      <RCTScrollView
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        initialNumToRender={0}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        testID="all-apps-letter-index-list"
        viewabilityConfigCallbackPairs={[]}
      >
        <View />
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaProvider>
`;

exports[`<AllAppsLetterIndex /> Tests should render correctly and match snapshot - populated list 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "bottom": 0,
          "flexDirection": "row",
          "justifyContent": "center",
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="all-apps-letter-index-wrapper"
    >
      <RCTScrollView
        data={
          [
            {
              "index": 0,
              "letter": "C",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        initialNumToRender={1}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={[]}
        testID="all-apps-letter-index-list"
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "height": 25,
                  "justifyContent": "center",
                  "width": 25,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#FFFFFF",
                    "fontSize": 12,
                    "fontWeight": "600",
                    "textShadowColor": "rgba(0, 0, 0, 0.75)",
                    "textShadowRadius": 2.5,
                  }
                }
              >
                C
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
